### PR TITLE
fix: rebind guilds/channels of cached messages after receiving `GUILD_CREATE` events

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

Autosharded clients rebind guild/channel references of cached messages after reconnecting to avoid keeping references to old/outdated objects, as they don't clear/recreate the caches (unlike unsharded clients). This was previously done in the `READY` handler, which generally only receives guilds in the `unavailable` state.
These guilds are effectively empty, which resulted in all cached messages' channel references being rebound to `Object`s instead of proper channels; to solve this, references will now be rebound only once all `GUILD_CREATE` events were received.

## Checklist

- [x] If code changes were made, then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [x] This PR fixes an issue
- [ ] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/state.py [hunk 1]
```diff
@@ -1956,17 +1956,17 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         self.shard_ids: Union[List[int], range] = []
         self.shards_launched: asyncio.Event = asyncio.Event()
 
-    def _update_message_references(self) -> None:
-        # self._messages won't be None when this is called
-        for msg in self._messages:  # type: ignore
+    def _update_guild_channel_references(self) -> None:
+        messages: Sequence[Message] = self._messages or []
+        for msg in messages:
```
Comment (onerandomusername) [inline disnake/state.py:1961]: this is minor, but let's be a bit more clear:
```suggestion
        if not self._messages:
            return
        for msg in self._messages:
```

**Review by onerandomusername — COMMENTED** [review]
a minor comment, but not sure if its worth making the change

File: disnake/state.py (hunk 1 above)
Comment (shiftinv) [reply disnake/state.py]: sounds good, changed in 165907e1de83f11e952d8dddea7feea120d10740

**Review by onerandomusername — APPROVED** [review]
thanks, this should help with caching

**Pull request merged**